src/main/drivers/system.c

```diff
@@ -60,16 +60,60 @@ void cycleCounterInit(void)
     usTicks = clocks.SYSCLK_Frequency / 1000000;
 }
 
+#include "common/atomic.h"
+#include "debug.h"
+
+static volatile int sysTickRollover = 0;
+
 // SysTick
 void SysTick_Handler(void)
 {
-    sysTickUptime++;
+    ATOMIC_BLOCK(NVIC_PRIO_MAX) {
+        sysTickUptime++;
+        sysTickRollover = 0;
+        *(volatile uint32_t *)&(SysTick->CTRL);
+    }
+}
+
+uint32_t microsISR(void)
+{
+    register uint32_t ms, cycle_cnt;
+
+    ATOMIC_BLOCK(NVIC_PRIO_MAX) {
+        cycle_cnt = SysTick->VAL;
+
+        if (SysTick->CTRL & SysTick_CTRL_COUNTFLAG_Msk) {
+            // Update pending.
+            // Remember it for calls within the same rollover period
+            // (Will be cleared when serviced).
+
+            // XXX Do we care for multiple rollovers?
+            // sysTickUptime would be incorrect in that case...
```

Comment:
It may be possible to handle `sysTickRollover` in systick handler ... but disabling interrupts for 1ms is terrible idea and wrong `sysTickUptime` will be least of the problems